File: mex/sources/korderpert/Dyn_pp/kord/first_order.cpp
/*1:*/


#include "kord_exception.h"
#include "first_order.h"
#include "cpplapack.h"

/*2:*/

int order_eigs(const double*alphar,const double*alphai,const double*beta)
{
	return(*alphar**alphar+*alphai**alphai<*beta**beta);
}


/*:2*/
;
/*3:*/

void FirstOrder::solve(const TwoDMatrix&fd)
{
	JournalRecordPair pa(journal);
	pa<<"Recovering first order derivatives "<<endrec;
	
	/*4:*/
	
	/*5:*/
	
	int off= 0;
	ConstTwoDMatrix fyplus(fd,off,ypart.nyss());
	off+= ypart.nyss();
	ConstTwoDMatrix fyszero(fd,off,ypart.nstat);
	off+= ypart.nstat;
	ConstTwoDMatrix fypzero(fd,off,ypart.npred);
	off+= ypart.npred;
	ConstTwoDMatrix fybzero(fd,off,ypart.nboth);
	off+= ypart.nboth;
	ConstTwoDMatrix fyfzero(fd,off,ypart.nforw);
	off+= ypart.nforw;
	ConstTwoDMatrix fymins(fd,off,ypart.nys());
	off+= ypart.nys();
	ConstTwoDMatrix fuzero(fd,off,nu);
	off+= nu;
	
	/*:5*/
	;
	/*6:*/
	
	int n= ypart.ny()+ypart.nboth;
	TwoDMatrix matD(n,n);
	matD.zeros();
	matD.place(fypzero,0,0);
	matD.place(fybzero,0,ypart.npred);
	matD.place(fyplus,0,ypart.nys()+ypart.nstat);
	for(int i= 0;i<ypart.nboth;i++)
		matD.get(ypart.ny()+i,ypart.npred+i)= 1.0;
	
	
	/*:6*/
	;
	/*7:*/
	
	TwoDMatrix matE(n,n);
	matE.zeros();
	matE.place(fymins,0,0);
	matE.place(fyszero,0,ypart.nys());
	matE.place(fyfzero,0,ypart.nys()+ypart.nstat+ypart.nboth);
	for(int i= 0;i<ypart.nboth;i++)
		matE.get(ypart.ny()+i,ypart.nys()+ypart.nstat+i)= -1.0;
	matE.mult(-1.0);
	
	/*:7*/
	;
	/*8:*/
	
	TwoDMatrix vsl(n,n);
	TwoDMatrix vsr(n,n);
	int lwork= 100*n+16;
	Vector work(lwork);
	IntSequence bwork(n);
	int info;
	LAPACK_dgges("N","V","S",order_eigs,&n,matE.getData().base(),&n,
		matD.getData().base(),&n,&sdim,alphar.base(),alphai.base(),
		beta.base(),vsl.getData().base(),&n,vsr.getData().base(),&n,
		work.base(),&lwork,&(bwork[0]),&info);
	bk_cond= (sdim==ypart.nys());
	
	
	/*:8*/
	;
	/*9:*/
	
	ConstGeneralMatrix z11(vsr,0,0,ypart.nys(),ypart.nys());
	ConstGeneralMatrix z12(vsr,0,ypart.nys(),ypart.nys(),n-ypart.nys());
	ConstGeneralMatrix z21(vsr,ypart.nys(),0,n-ypart.nys(),ypart.nys());
	ConstGeneralMatrix z22(vsr,ypart.nys(),ypart.nys(),n-ypart.nys(),n-ypart.nys());
	
	/*:9*/
	;
	/*10:*/
	
	GeneralMatrix sfder(z12,"transpose");
	z22.multInvLeftTrans(sfder);
	sfder.mult(-1);
	
	/*:10*/
	;
	/*11:*/
	
	ConstGeneralMatrix s11(matE,0,0,ypart.nys(),ypart.nys());
	ConstGeneralMatrix t11(matD,0,0,ypart.nys(),ypart.nys());
	GeneralMatrix dumm(s11,"transpose");
	z11.multInvLeftTrans(dumm);
	GeneralMatrix preder(dumm,"transpose");
	t11.multInvLeft(preder);
	preder.multLeft(z11);
	
	/*:11*/
	;
	/*12:*/
	
	gy.place(preder,ypart.nstat,0);
	GeneralMatrix sder(sfder,0,0,ypart.nstat,ypart.nys());
	gy.place(sder,0,0);
	GeneralMatrix fder(sfder,ypart.nstat+ypart.nboth,0,ypart.nforw,ypart.nys());
	gy.place(fder,ypart.nstat+ypart.nys(),0);
	
	/*:12*/
	;
	/*13:*/
	
	GeneralMatrix bder((const GeneralMatrix&)sfder,ypart.nstat,0,ypart.nboth,ypart.nys());
	GeneralMatrix bder2(preder,ypart.npred,0,ypart.nboth,ypart.nys());
	bder.add(-1,bder2);
	b_error= bder.getData().getMax();
	
	/*:13*/
	;
	
	
	/*:4*/
	;
	/*14:*/
	
	GeneralMatrix matA(ypart.ny(),ypart.ny());
	matA.zeros();
	ConstGeneralMatrix gss(gy,ypart.nstat+ypart.npred,0,ypart.nyss(),ypart.nys());
	GeneralMatrix aux(fyplus,gss);
	matA.place(aux,0,ypart.nstat);
	ConstGeneralMatrix fyzero(fd,0,ypart.nyss(),ypart.ny(),ypart.ny());
	matA.add(1.0,fyzero);
	gu.zeros();
	gu.add(-1.0,fuzero);
	ConstGeneralMatrix(matA).multInvLeft(gu);
	
	/*:14*/
	;
	journalEigs();
	
	if(!gy.isFinite()||!gu.isFinite()){
		throw KordException(__FILE__,__LINE__,
			"NaN or Inf asserted in first order derivatives in FirstOrder::solve");
	}
}

/*:3*/
;
/*15:*/

void FirstOrder::journalEigs()
{
	if(bk_cond){
		JournalRecord jr(journal);
		jr<<"Blanchard-Kahn conditition satisfied, model stable"<<endrec;
	}else{
		JournalRecord jr(journal);
		jr<<"Blanchard-Kahn condition not satisfied, model not stable: sdim="<<sdim
			<<" "<<"npred="<<ypart.nys()<<endrec;
	}
	if(!bk_cond){
		for(int i= 0;i<alphar.length();i++){
			if(i==sdim||i==ypart.nys()){
				JournalRecord jr(journal);
				jr<<"---------------------------------------------------- ";
				if(i==sdim)
					jr<<"sdim";
				else
					jr<<"npred";
				jr<<endrec;
			}
			JournalRecord jr(journal);
			double mod= sqrt(alphar[i]*alphar[i]+alphai[i]*alphai[i]);
			mod= mod/round(100000*std::abs(beta[i]))*100000;
			jr<<i<<"\t("<<alphar[i]<<","<<alphai[i]<<") / "<<beta[i]
				<<"  \t"<<mod<<endrec;
		}
	}
}


/*:15*/
;

/*:1*/

File: mex/sources/k_order_perturbation/tests/first_order.m
function [gy]=first_order(M_, dr, jacobia)
% Emulation of Dynare++ c++ first_order.cpp for testing pruposes

% Copyright (C) 2009 Dynare Team
%
% This file is part of Dynare.
%
% Dynare is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% Dynare is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with Dynare.  If not, see <http://www.gnu.org/licenses/>.

% fd = jacobia_
% reorder jacobia_
[fd]=k_reOrderedJacobia(M_, jacobia)

%ypart=dr;
ypart.ny=M_.endo_nbr;
ypart.nyss=dr.nfwrd+dr.nboth;
ypart.nys=dr.npred;
ypart.npred=dr.npred-dr.nboth;
ypart.nboth=dr.nboth;
ypart.nforw=dr.nfwrd;
ypart.nstat =dr.nstatic
nu=M_.exo_nbr

off= 1; % = 0 in C
fyplus = fd(:,off:off+ypart.nyss-1);
off= off+ypart.nyss;
fyszero= fd(:,off:off+ypart.nstat-1);
off= off+ypart.nstat;
fypzero= fd(:,off:off+ypart.npred-1);
off= off+ypart.npred;
fybzero= fd(:,off:off+ypart.nboth-1);
off= off+ypart.nboth;
fyfzero= fd(:,off:off+ypart.nforw-1);
off= off+ypart.nforw;
fymins= fd(:,off:off+ypart.nys-1);
off= off+ypart.nys;
fuzero= fd(:,off:off+nu-1);
off=off+ nu;

n= ypart.ny+ypart.nboth;
%TwoDMatrix 
matD=zeros(n,n);
%	matD.place(fypzero,0,0);
matD(1:n-ypart.nboth,1:ypart.npred)= fypzero;
%	matD.place(fybzero,0,ypart.npred);
matD(1:n-ypart.nboth,ypart.npred+1:ypart.npred+ypart.nboth)=fybzero;
%	matD.place(fyplus,0,ypart.nys()+ypart.nstat);
matD(1:n-ypart.nboth,ypart.nys+ypart.nstat+1:ypart.nys+ypart.nstat+ypart.nyss)=fyplus;
for i=1:ypart.nboth
    matD(ypart.ny()+i,ypart.npred+i)= 1.0;
end

matE=[fymins, fyszero, zeros(n-ypart.nboth,ypart.nboth), fyfzero; zeros(ypart.nboth,n)];
%	matE.place(fymins;
%	matE.place(fyszero,0,ypart.nys());
%	matE.place(fyfzero,0,ypart.nys()+ypart.nstat+ypart.nboth);

for i= 1:ypart.nboth
    matE(ypart.ny()+i,ypart.nys()+ypart.nstat+i)= -1.0;
end
matE=-matE; %matE.mult(-1.0);

%    vsl=zeros(n,n);
%	vsr=zeros(n,n);
%	lwork= 100*n+16;
%	work=zeros(1,lwork);
%	bwork=zeros(1,n);
%int info;

%    	LAPACK_dgges("N","V","S",order_eigs,&n,matE.getData().base(),&n,
%		matD.getData().base(),&n,&sdim,alphar.base(),alphai.base(),
%		beta.base(),vsl.getData().base(),&n,vsr.getData().base(),&n,
%		work.base(),&lwork,&(bwork[0]),&info);

[matE1,matD1,vsr,sdim,dr.eigval,info] = mjdgges(matE,matD,1);

bk_cond= (sdim==ypart.nys);

%  	ConstGeneralMatrix z11(vsr,0,0,ypart.nys(),ypart.nys());
z11=vsr(1:ypart.nys,1:ypart.nys);
%	ConstGeneralMatrix z12(vsr,0,ypart.nys(),ypart.nys(),n-ypart.nys());
z12=vsr(1:ypart.nys(),ypart.nys+1:end);%, n-ypart.nys);
                                       %	ConstGeneralMatrix z21(vsr,ypart.nys(),0,n-ypart.nys(),ypart.nys());
z21=vsr(ypart.nys+1:end,1:ypart.nys);
%	ConstGeneralMatrix z22(vsr,ypart.nys(),ypart.nys(),n-ypart.nys(),n-ypart.nys());
z22=vsr(ypart.nys+1:end,ypart.nys+1:end);

% 	GeneralMatrix sfder(z12,"transpose");
sfder=z12';%,"transpose");
           %	z22.multInvLeftTrans(sfder);
sfder=z22'\sfder;
sfder=-sfder;% .mult(-1);

%s11(matE,0,0,ypart.nys(),ypart.nys());
s11=matE1(1:ypart.nys,1:ypart.nys);
%	 t11=(matD1,0,0,ypart.nys(),ypart.nys());
t11=matD1(1:ypart.nys,1:ypart.nys);
dumm=(s11');%,"transpose");
            %z11.multInvLeftTrans(dumm);
dumm=z11'\dumm;
preder=(dumm');%,"transpose");
               %t11.multInvLeft(preder);
preder=t11\preder;
%preder.multLeft(z11);
preder= z11*preder;

%	gy.place(preder,ypart.nstat,0);
%	gy=(zeros(ypart.nstat,size(preder,2)) ;preder);
%	 sder(sfder,0,0,ypart.nstat,ypart.nys());
sder=sfder(1:ypart.nstat,1:ypart.nys);
%	gy.place(sder,0,0);
%	gy(1:ypart.nstat, 1:ypart.nys)=sder;
%    gy=[sder;preder];
%	 fder(sfder,ypart.nstat+ypart.nboth,0,ypart.nforw,ypart.nys());
fder=sfder(ypart.nstat+ypart.nboth+1:ypart.nstat+ypart.nboth+ypart.nforw,1:ypart.nys);
%	gy.place(fder,ypart.nstat+ypart.nys(),0);
%	gy(ypart.nstat+ypart.nys,1)=fder;
gy=[sder;preder;fder];
